draft-ietf-quic-tls.md

```diff
@@ -1007,144 +813,94 @@ encrypted_pn = ChaCha20(pn_key, counter, nonce, packet_number)
 
-A server might choose to retain 0-RTT packets that arrive before a TLS
-ClientHello.  The server can then use those packets once the ClientHello
-arrives.  However, the potential for denial of service from buffering 0-RTT
-packets is significant.  These packets cannot be authenticated and so might be
-employed by an attacker to exhaust server resources.  Limiting the number of
-packets that are saved might be necessary.
+A server MUST NOT use 0-RTT keys to protect packets.
 
-The server transitions to using 1-RTT keys after sending its first flight of TLS
-handshake messages, ending in the Finished.
-From this point, the server protects all packets with 1-RTT
-keys.  Future packets are therefore protected with 1-RTT keys.  Initially, these
-are marked with a KEY_PHASE of 0.
+If a server rejects 0-RTT, then the TLS stream will not include any TLS records
+protected with 0-RTT keys.
 
+: 0-RTT data can be acknowledged by the server as it receives it, but any
+  packets containing acknowledgments of 0-RTT data cannot have packet protection
+  removed by the client until the TLS handshake is complete.  The 1-RTT keys
+  necessary to remove packet protection cannot be derived until the client
+  receives all server handshake messages.
 
-### Retransmission and Acknowledgment of Unprotected Packets
 
-TLS handshake messages from both client and server are critical to the key
-exchange.  The contents of these messages determine the keys used to protect
-later messages.  If these handshake messages are included in packets that are
-protected with these keys, they will be indecipherable to the recipient.
+## Receiving Out-of-Order Protected Frames {#pre-hs-protected}
 
-Even though newer keys could be available when retransmitting, retransmissions
-of these handshake messages MUST be sent in packets protected with handshake
-keys.  An endpoint MUST generate ACK frames for these messages and send them in
-packets protected with handshake keys.
+Due to reordering and loss, protected packets might be received by an endpoint
+before the final TLS handshake messages are received.  A client will be unable
+to decrypt 1-RTT packets from the server, whereas a server will be able to
+decrypt 1-RTT packets from the client.
 
-A HelloRetryRequest handshake message might be used to reject an initial
-ClientHello.  A HelloRetryRequest handshake message is sent in a Retry packet;
-any second ClientHello that is sent in response uses a Initial packet type.
-These packets are only protected with a predictable key (see
-{{handshake-secrets}}).  This is natural, because no shared secret will be
-available when these messages need to be sent.  Upon receipt of a
-HelloRetryRequest, a client SHOULD cease any transmission of 0-RTT data; 0-RTT
-data will only be discarded by any server that sends a HelloRetryRequest.
+Packets protected with 1-RTT keys MAY be stored and later decrypted and used
+once the handshake is complete.  A server MUST NOT use 1-RTT protected packets
```

Comment:
As above, I'd be more specific than "use" -- the server isn't permitted to remove protection from 1-RTT packets, but it can use the keys to protect 1-RTT packets on the outbound path.

Reply:
Adjusted some.